### PR TITLE
drivers: caam Fixes for Era 10 and above

### PR DESCRIPTION
**Pull request by ruchi393**

### CODE REVIEW

**Comment by ruchi393** [comment]
@clementfaure , please help review.

File: core/arch/arm/plat-ls/conf.mk [hunk 1]
```diff
@@ -146,7 +147,7 @@ ifeq ($(CFG_NXP_CAAM),y)
 # it with generic crypto API can be enabled.
 CFG_CRYPTO_DRIVER ?= y
 CFG_CAAM_64BIT ?= y
-CFG_CRYPTO_DRIVER_DEBUG ?= n
+CFG_CRYPTO_DRIVER_DEBUG ?= y
```
Comment (sahilnxp) [inline core/arch/arm/plat-ls/conf.mk:150]: Please keep CFG_CRYPTO_DRIVER_DEBUG = n by default.

File: core/arch/arm/plat-ls/crypto_conf.mk [hunk 2]
```diff
@@ -30,18 +34,23 @@ ifneq (,$(filter $(PLATFORM_FLAVOR),ls1046ardb))
...
 #
-ifeq ($(CFG_CRYPTO_DRIVER),y)
+ifeq ($(CFG_CRYPTO_DRIVER), y)
 
-$(call force,CFG_NXP_CAAM_RUNTIME_JR,y)
+$(call force, CFG_NXP_CAAM_RUNTIME_JR, y)
 
 #
-# Definition of all HW accelerations for all LS
+# Definition of all HW accelerations for all i.MX
```
Comment (sahilnxp) [inline core/arch/arm/plat-ls/crypto_conf.mk:50]: Please keep this to LS.

File: core/arch/arm/plat-ls/conf.mk (hunk 1 above)
Comment (ruchi393) [reply core/arch/arm/plat-ls/conf.mk:150]: Ok

**Comment by ruchi393** [comment]
Rebased the branch and took care of comments by @sahilnxp

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c [hunk 3]
```diff
@@ -10,50 +10,82 @@
 #include <registers/ctrl_regs.h>
 #include <registers/version_regs.h>
 
-uint8_t caam_hal_ctrl_jrnum(vaddr_t baseaddr)
+uint8_t caam_hal_ctrl_era(vaddr_t baseaddr)
 {
 	uint32_t val = 0;
```
Comment (clementfaure) [inline core/drivers/crypto/caam/hal/common/hal_ctrl.c:15]: ```uint32_t val = io_caam_read32(baseaddr + CCBVID);```

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c [hunk 4]
```diff
@@ -10,50 +10,82 @@
...
 
-uint8_t caam_hal_ctrl_hash_limit(vaddr_t baseaddr)
+uint8_t caam_hal_ctrl_jrnum(vaddr_t baseaddr)
 {
 	uint32_t val = 0;
+	uint8_t jrnum;
 
-	/* Read the number of instance */
-	val = io_caam_read32(baseaddr + CHANUM_LS);
+	if (caam_hal_ctrl_era(baseaddr) < 10) {
```
Comment (clementfaure) [inline core/drivers/crypto/caam/hal/common/hal_ctrl.c:28]: What is 10? A version number?

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c [hunk 5]
```diff
@@ -10,50 +10,82 @@
...
+		jrnum = GET_CHANUM_MS_JRNUM(val);
+	} else {
+		val = io_caam_read32(baseaddr + JR_VERSION);
 
-		return TEE_MAIN_ALGO_SHA512;
+		jrnum = GET_JR_VERSION_JRNUM(val);
 	}
 
-	return UINT8_MAX;
+	return jrnum;
```
Comment (clementfaure) [inline core/drivers/crypto/caam/hal/common/hal_ctrl.c:38]: You could do ```return GET_CHANUM_MS_JRNUM(val);``` and ```return GET_JR_VERSION_JRNUM(val);``` so you don't need ```uint8_t jrnum;```
Up to you

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c [hunk 6]
```diff
@@ -10,50 +10,82 @@
...
 }
 
-uint8_t caam_hal_ctrl_pknum(vaddr_t baseaddr)
+uint8_t caam_hal_ctrl_hash_limit(vaddr_t baseaddr)
 {
 	uint32_t val = 0;
 
-	val = io_caam_read32(baseaddr + CHANUM_LS);
+	if (caam_hal_ctrl_era(baseaddr) < 10) {
```
Comment (clementfaure) [inline core/drivers/crypto/caam/hal/common/hal_ctrl.c:45]: What is 10 ?

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c [hunk 7]
```diff
@@ -10,50 +10,82 @@
...
+	}
+
+	return UINT8_MAX;
 }
 
-uint8_t caam_hal_ctrl_era(vaddr_t baseaddr)
+uint8_t caam_hal_ctrl_pknum(vaddr_t baseaddr)
 {
 	uint32_t val = 0;
+	uint8_t pknum;
```
Comment (clementfaure) [inline core/drivers/crypto/caam/hal/common/hal_ctrl.c:78]: Please initialize.

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c [hunk 8]
```diff
@@ -10,50 +10,82 @@
...
 
-	/* Read the number of instance */
-	val = io_caam_read32(baseaddr + CCBVID);
+	if (caam_hal_ctrl_era(baseaddr) < 10) {
```
Comment (clementfaure) [inline core/drivers/crypto/caam/hal/common/hal_ctrl.c:80]: What is 10?

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c [hunk 9]
```diff
@@ -10,50 +10,82 @@
...
+		val = io_caam_read32(baseaddr + CHANUM_LS);
 
-	return GET_CCBVID_CAAM_ERA(val);
+		pknum = GET_CHANUM_LS_PKNUM(val);
+	} else {
+		val = io_caam_read32(baseaddr + PKHA_VERSION);
+
```
Comment (clementfaure) [inline core/drivers/crypto/caam/hal/common/hal_ctrl.c:86]: Delete empty line

File: core/drivers/crypto/caam/hal/common/hal_rng.c [hunk 10]
```diff
@@ -6,22 +6,31 @@
  *         Implementation of primitives to access HW.
  */
 #include <caam_hal_rng.h>
+#include <caam_hal_ctrl.h>
```
Comment (clementfaure) [inline core/drivers/crypto/caam/hal/common/hal_rng.c:9]: Please sort headers

**Comment by sahilnxp** [comment]
`Reviewed-by: Sahil Malhotra <sahil.malhotra@nxp.com>`

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 6 above)
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:45]: 10 is the SEC Era

File: core/arch/arm/plat-ls/conf.mk [hunk 11]
```diff
@@ -146,7 +147,7 @@ ifeq ($(CFG_NXP_CAAM),y)
 # it with generic crypto API can be enabled.
 CFG_CRYPTO_DRIVER ?= y
 CFG_CAAM_64BIT ?= y
-CFG_CRYPTO_DRIVER_DEBUG ?= n
+CFG_CRYPTO_DRIVER_DEBUG ?= 0
```
[the remark this answers is not in the document]
Comment (ruchi393) [reply core/arch/arm/plat-ls/conf.mk:150]: Ok

File: core/drivers/crypto/caam/hal/common/registers/version_regs.h [hunk 12]
```diff
@@ -43,5 +43,28 @@
 #define BM_CHANUM_LS_MDNUM              SHIFT_U32(0xF, 12)
 #define GET_CHANUM_LS_MDNUM(val)	(((val) & BM_CHANUM_LS_MDNUM) >> 12)
 
+/* PKHA Version for Era > 10 */
+#define PKHA_VERSION			0x0E8C
+#define BM_PKHA_VERSION_PKNUM           0xFF
```
[the remark this answers is not in the document]
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/registers/version_regs.h:48]: Its already aligned. I am not sure why it is showing like this in the patch

File: core/drivers/crypto/caam/hal/common/registers/version_regs.h [hunk 13]
```diff
@@ -43,5 +43,28 @@
...
+#define MDHA_VERSION_MDVID_LP256	SHIFT_U32(0, 24)
+
+/* RNG Version for Era > 10 */
+#define RNG_VERSION			0x0EF8
+#define BM_RNG_VERSION_VID	        SHIFT_U32(0xFF, 24)
+#define GET_RNG_VERSION_VID(val)	(((val) & BM_RNG_VERSION_VID))
+
+/* JR Version for Era > 10 */
+#define JR_VERSION			0x0EF8
+#define BM_JR_VERSION_JRNUM             0xFF
```
[the remark this answers is not in the document]
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/registers/version_regs.h:66]: Its already aligned. I am not sure why it is showing like this in the patch

File: core/arch/arm/plat-ls/crypto_conf.mk (hunk 2 above)
[the remark this answers is not in the document]
Comment (ruchi393) [reply core/arch/arm/plat-ls/crypto_conf.mk]: ok

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 3 above)
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:15]: Sure, I will change that

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 4 above)
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:28]: 10 is SEC Era Number

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 8 above)
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:80]: CAAM Era

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 9 above)
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:86]: I had added the blank line for clarity but I will remove it

File: core/drivers/crypto/caam/hal/common/hal_rng.c (hunk 10 above)
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/hal_rng.c:9]: ok

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 5 above)
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:38]: I did that earlier and then changes it..I will revert to return from if else loop and avoid the new variable

File: core/drivers/crypto/caam/hal/common/registers/version_regs.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/registers/version_regs.h:66]: I got it, instead of tabs, spaces were used. I will correct that.

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 6 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:38]: Ok

**Comment by clementfaure** [comment]
Please fix the last error reported by the CI and it should be good.
```Reviewed-by: Clement Faure <clement.faure@nxp.com>```

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 7 above)
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c]: done

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 5 above)
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:34]: changing it to return gives checkpatch error, so I have reverted it to variable approach.

Comment (clementfaure) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:34]: Checkpatch just asked to remove the ```else``` condition that was unnecessary since there was a ```return``` instruction right before. 
It does not really matter anyway.

**Comment by jforissier** [comment]
Just a minor thing, the crash dump in commit "drivers: caam: Fix alignment fault caused by caam_desc_pop()" would be more useful with debug info (`symbolize.py`). I mean the context of the problem would be more clear.

File: core/arch/arm/plat-ls/crypto_conf.mk [hunk 14]
```diff
@@ -3,20 +3,24 @@
...
+# DBG_CTRL   BIT32(1)  // Controller trace
+# DBG_MEM    BIT32(2)  // Memory utility trace
+# DBG_SGT    BIT32(3)  // Scatter Gather trace
+# DBG_PWR    BIT32(4)  // Power trace
+# DBG_JR     BIT32(5)  // Job Ring trace
+# DBG_RNG    BIT32(6)  // RNG trace
+# DBG_HASH   BIT32(7)  // Hash trace
+# DBG_RSA    BIT32(8)  // RSA trace
+# DBG_CIPHER BIT32(9)  // Cipher trace
+CFG_DBG_CAAM_TRACE ?= 0x2
```
Comment (etienne-lms) [inline core/arch/arm/plat-ls/crypto_conf.mk:21]: `BIT(1)` ?

File: core/arch/arm/plat-ls/crypto_conf.mk [hunk 15]
```diff
@@ -30,18 +34,23 @@ ifneq (,$(filter $(PLATFORM_FLAVOR),ls1046ardb))
 $(call force,CFG_JR_INT,137)  # Default JR IT Number (105 + 32) = 137
 endif
 
+ifneq (,$(filter $(PLATFORM_FLAVOR),lx2160ardb))
+$(call force,CFG_JR_INT,174)
+$(call force,CFG_NB_JOBS_QUEUE,80)  # Default JR index used
```
Comment (etienne-lms) [inline core/arch/arm/plat-ls/crypto_conf.mk:39]: with or without a space char?

File: core/arch/arm/plat-ls/crypto_conf.mk [hunk 16]
```diff
@@ -30,18 +34,23 @@ ifneq (,$(filter $(PLATFORM_FLAVOR),ls1046ardb))
...
+endif
+
 #
 # Configuration of the Crypto Driver
 #
-ifeq ($(CFG_CRYPTO_DRIVER),y)
+ifeq ($(CFG_CRYPTO_DRIVER), y)
```
Comment (etienne-lms) [inline core/arch/arm/plat-ls/crypto_conf.mk:45]: unrelated change. needed?

File: core/arch/arm/plat-ls/crypto_conf.mk [hunk 17]
```diff
@@ -51,11 +60,32 @@ $(call force,CFG_WITH_SOFTWARE_PRNG,n)
...
+# CFG_CRYPTO_xxx_HW is 'y'
+cryphw-one-enabled = $(call cfg-one-enabled, \
+                        $(foreach v,$(1), CFG_NXP_CAAM_$(v)_DRV))
+
 # Definition of the HW and Cryto Driver Algorithm supported by all LS
-$(eval $(call cryphw-enable-drv-hw,HASH))
+$(eval $(call cryphw-enable-drv-hw, HASH))
+$(eval $(call cryphw-enable-drv-hw, CIPHER))
+
+$(eval $(call cryphw-enable-drv-hw, RSA))
```
Comment (etienne-lms) [inline core/arch/arm/plat-ls/crypto_conf.mk:76]: remove empty line?

File: core/arch/arm/plat-ls/crypto_conf.mk [hunk 18]
```diff
@@ -51,11 +60,32 @@ $(call force,CFG_WITH_SOFTWARE_PRNG,n)
...
+
+# Define the RSA Private Key Format used by the CAAM
+#   Format #1: (n, d)
+#   Format #2: (p, q, d)
+#   Format #3: (p, q, dp, dq, qp)
+CFG_NXP_CAAM_RSA_KEY_FORMAT ?= 3
+
+$(call force, CFG_NXP_CAAM_ACIPHER_DRV, $(call cryphw-one-enabled, RSA))
```
Comment (etienne-lms) [inline core/arch/arm/plat-ls/crypto_conf.mk:84]: could remove empty space.

File: core/arch/arm/plat-ls/crypto_conf.mk (hunk 14 above)
Comment (ruchi393) [reply core/arch/arm/plat-ls/crypto_conf.mk:21]: I will change it to BIT32(1)

File: core/arch/arm/plat-ls/crypto_conf.mk (hunk 16 above)
Comment (ruchi393) [reply core/arch/arm/plat-ls/crypto_conf.mk:45]: No, will remove.

File: core/arch/arm/plat-ls/crypto_conf.mk (hunk 17 above)
Comment (ruchi393) [reply core/arch/arm/plat-ls/crypto_conf.mk:76]: Ok

File: core/arch/arm/plat-ls/crypto_conf.mk (hunk 18 above)
Comment (ruchi393) [reply core/arch/arm/plat-ls/crypto_conf.mk:84]: Ok

**Comment by etienne-lms** [comment]
Coudl you squash commit "crypto_conf.mk: Review comments incorporated" into commit "core: plat-ls: Enable CAAM driver for PLATFORM lx2160ardb"

For commit "drivers: caam: Add register map changes for Era 10"
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

For commit "drivers: caam: Fix alignment fault caused by caam_desc_pop()"
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` with comment (empty line) addressed.

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 6 above)
Comment (etienne-lms) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:41]: As this value 10 is used at several places, I suggest to use a macro, i.e. `CAAM_SEC_ERA`.

File: core/drivers/crypto/caam/hal/common/registers/version_regs.h [hunk 19]
```diff
@@ -43,5 +43,28 @@
 #define BM_CHANUM_LS_MDNUM              SHIFT_U32(0xF, 12)
 #define GET_CHANUM_LS_MDNUM(val)	(((val) & BM_CHANUM_LS_MDNUM) >> 12)
 
+/* PKHA Version for Era > 10 */
+#define PKHA_VERSION			0x0E8C
+#define BM_PKHA_VERSION_PKNUM		0xFF
+#define GET_PKHA_VERSION_PKNUM(val)	(((val) & BM_PKHA_VERSION_PKNUM))
```
Comment (etienne-lms) [inline core/drivers/crypto/caam/hal/common/registers/version_regs.h:49]: can remove outer parentheses.
3 other occurrences below.

File: core/drivers/crypto/caam/caam_desc.c [hunk 20]
```diff
@@ -80,12 +80,14 @@ void caam_desc_push(struct caam_inring_entry *in_entry, paddr_t paddr)
 
 paddr_t caam_desc_pop(struct caam_outring_entry *out_entry)
 {
+	const uint32_t *a32 = (const uint32_t *)(&out_entry->desc);
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/crypto/caam/caam_desc.c:83]: ping!

[the remark this answers is not in the document]
Comment (ruchi393) [reply core/drivers/crypto/caam/caam_desc.c:83]: Looks like I missed the empty line during rework. Will fix it. However this change is in commit "drivers: caam: Fix alignment fault caused by caam_desc_pop()". I will fix it in that

File: core/drivers/crypto/caam/hal/common/registers/version_regs.h (hunk 19 above)
Comment (ruchi393) [reply core/drivers/crypto/caam/hal/common/registers/version_regs.h:49]: Ok
